src/Heading/Heading.tsx

```diff
@@ -0,0 +1,81 @@
+import { ElementType, FC } from 'react';
+import { SetOptional } from 'type-fest';
+import styled from '@emotion/styled';
+import { css, Theme } from '@emotion/react';
+
+import { FontSize, FontWeight } from '../theme/typography';
+import Text from '../Text';
+
+const lineHeightMultiplier = 0.5;
+
+const baseStyles = ({ theme }: { theme: Theme }) => css`
+  line-height: ${theme.lineHeights.small};
+`;
+
+const calculatedMargin = ({
+  fontSize,
+  mb,
+  theme
+}: {
+  fontSize: FontSize;
+  mb: string;
+  theme: Theme;
+}) =>
+  !mb &&
+  css`
+    margin-bottom: calc(${theme.fontSizes[fontSize]} * ${lineHeightMultiplier});
+  `;
+
+type Props = {
+  fontSize: FontSize;
+  fontWeight?: FontWeight;
+  mb: string;
+};
+
+const Heading = styled(Text)<Props>`
+  ${baseStyles};
+  ${calculatedMargin};
+`;
+
+Heading.defaultProps = {
+  as: 'h1',
+  color: 'text.dark',
+  fontWeight: 'bold',
+  lineHeight: 'small'
+};
+
+const createHeading =
+  (
+    as: ElementType,
+    defaultFontSize: FontSize
+  ): FC<SetOptional<Props, 'fontSize'>> =>
+  ({ children, fontSize = defaultFontSize, ...props }) =>
+    (
+      <Heading as={as} fontSize={fontSize} {...props}>
+        {children}
+      </Heading>
+    );
+
+const H1 = createHeading('h1', 'size7');
+
+const H2 = createHeading('h2', 'size6');
+
+const H3 = createHeading('h3', 'size5');
+
+const H4 = createHeading('h4', 'size4');
+
+const H5 = styled(createHeading('h5', 'size4'))`
+  line-height: ${(props) => props.theme.lineHeights.large};
+`;
+H5.defaultProps = {
+  fontWeight: 'medium'
+};
+
+const H6 = styled(createHeading('h6', 'size4'))`
+  line-height: ${(props) => props.theme.lineHeights.large};
+`;
+H6.defaultProps = {
+  fontWeight: 'regular'
+};
+
+export default Object.assign(Heading, { H1, H2, H3, H4, H5, H6 });
```

Comment:
This gives us the same structure we had before, with `H1`, `H2`, etc. attached to `Heading` as properties.  This is much less awkward than trying to type `Heading` explicitly imo.

Incidentally, this would probably work for attaching `fragments` to components in the main app if we wanted to go back to that, but I haven't tested it.

Reply:
It seems like this runs the risk eventually of Typescript realizing what's happening here, and making the same complaint that it does if you assign the fields directly.

Reply:
We can definitely migrate away from it if it's a concern.  I found other TS/React libraries doing this FWIW, e.g. https://github.com/react-bootstrap/react-bootstrap/blob/master/src/Dropdown.tsx#L246-L253.

I'll look more into the implications when I have a chance anyway, because you raise an interesting point.